### PR TITLE
Fix activity opens to a black canvas and key errors

### LINKED ISSUE #21: Activity opens to black canvas
Activity currently opens to a black canvas.

### PR DESCRIPTION
**Pull request by vaibhav-sangwan**
VIDEORESIZE event is triggered on activity startup, which is handled by calling pygame.display.set_mode(). This returns an empty Surface object completely discarding the previous state of screen.

* Call draw functions after resizing to populate the newly returned empty surface with game elements,

* Initialize pygame mixer.

Fixes #21 , and the potential issues encountered in #22 (of not being able to resize the window) 
@chimosky, @quozl, @walterbender please review.

### CODE REVIEW

**Comment by walterbender** [comment]
I am not able to test because I get stuck on a key error when trying to get past the landing page.
It is not very Sugary to use a landing page for starting/settings/quit in any case. But the landing page is visible with your patch.

**Comment by vaibhav-sangwan** [comment]
I've updated Sugargame so that key events can be correctly registered. However, The activity gets stuck on the landing page after pressing "C". The following is error is shown in logs -
```
Traceback (most recent call last):
 File "/home/vishu/Activities/ball-and-brick-activity/game.py", line 601, in run
   settings()
 File "/home/vishu/Activities/ball-and-brick-activity/game.py", line 471, in settings
   gameLoop()
 File "/home/vishu/Activities/ball-and-brick-activity/game.py", line 512, in gameLoop
   check_input()
 File "/home/vishu/Activities/ball-and-brick-activity/game.py", line 209, in check_input
   if keys[pygame.K_LEFT]:
IndexError: list index out of range
```

This key error isn't encountered when game.py is executed from main, suggesting that there's an issue with Sugargame. 
While debugging, I noticed that on line 207 of game.py, data type of "keys" is ScanCodeWrapper when the file is executed from main. Whereas the type of "keys" is list when it's run from sugar activity ring.

**Comment by vaibhav-sangwan** [comment]
The key events should work fine now.

**Comment by quozl** [comment]
pygame.mixer should be in modules parameter of the call to PygameCanvas, which is also missing pygame.font.

Before your commits, current _master_ branch does not run on Ubuntu 20.04 with Pygame 1.9.6, failing with mixer not initialised.

File: game.py [hunk 1]
```diff
@@ -487,8 +508,13 @@ def gameLoop():
...
             pygame.display.update()
             restart_game()
+            self.keys_pressed = {
+                pygame.K_LEFT: False,
+                pygame.K_RIGHT: False
+            }
             while 1:
-                for event in pygame.event.get():
+                self.py_events = pygame.event.get()
+                for event in self.py_events:
```
Comment (quozl) [inline game.py:517]: Why this change?  Storing the events and reprocessing them?  Why not process them here?

Comment (vaibhav-sangwan) [reply game.py:517]: pygame.event.get() removes the events from the event queue ([documentation](https://www.pygame.org/docs/ref/event.html#pygame.event.get)). The key events are to be handled in check_inputs() function, whereas quit and pause check events were being handled in the gameLoop() itself.

**Comment by vaibhav-sangwan** [comment]
> Before your commits, current _master_ branch does not run on Ubuntu 20.04 with Pygame 1.9.6, failing with mixer not initialised.

Tested current _master_ branch on Ubuntu 22.04 with pygame 2.5.2 (Python 3.10.12). Activity fails to open with the same error:-
```
Traceback (most recent call last):
 File "/usr/bin/sugar-activity3", line 5, in <module>
   activityinstance.main()
 File "/usr/lib/python3/dist-packages/sugar3/activity/activityinstance.py", line 230, in main
   instance = create_activity_instance(activity_constructor, activity_handle)
 File "/usr/lib/python3/dist-packages/sugar3/activity/activityinstance.py", line 59, in create_activity_instance
   activity = constructor(handle)
 File "/home/vishu/Activities/ball-and-brick-activity/activity.py", line 32, in __init__
   self.game = game.BallAndBrick()
 File "/home/vishu/Activities/ball-and-brick-activity/game.py", line 94, in __init__
   self.brick_hit_sound  = pygame.mixer.Sound("assets/brickhit.ogg")
pygame.error: mixer not initialized
```

**Comment by walterbender** [comment]
It works. Probably it should under a separate MR, but refactoring the command to be toolbar buttons would make it much nicer.

File: game.py (hunk 1 above)
Comment (chimosky) [reply game.py:516]: The event queue needs to be cleared so it doesn't get filled, are you saying some events aren't being handled?

File: game.py [hunk 2]
```diff
@@ -106,6 +101,10 @@ def read_file(self, file_path):
         pass
 
     def run(self):
+        self.brick_hit_sound  = pygame.mixer.Sound("assets/brickhit.ogg")
+        self.paddle_hit_sound = pygame.mixer.Sound("assets/paddlehit.ogg")
+        self.brick_hit_sound.set_volume(0.8)
+
```
Comment (chimosky) [inline game.py:107]: Why move these here?

Comment (vaibhav-sangwan) [reply game.py:107]: In activity.py, a new object of BallAndBrick class is created before the call to PygameCanvas constructor (where all the modules are initialized). So the code inside \_\_init__() would run before any modules are initialized resulting in `mixer not initialized` error. Moving these lines to the run function solves this problem.

File: game.py (hunk 1 above)
Comment (vaibhav-sangwan) [reply game.py:516]: Yes, the pause event wasn't being handled correctly and the movements were janky when I tried to use pygame.event.get() in both _gameLoop_ and  _checkInputs_ functions. Storing the events in self.py_events solves this issue and also clears up the event queue.

Another alternative was to handle all the events in one place (in the gameLoop itself), but I chose to keep the checkInputs() function to keep the code more modular.

**Comment by walterbender** [comment]
> > refactoring the command to be toolbar buttons would make it much nicer.
> 
> Could you please clarify a bit more on what command you are referring to?

Buttons for play, settings, quit. Right now those are keyboard inputs from the landing page.

**Comment by vaibhav-sangwan** [comment]
Yes, that would be more user-friendly. I'll add title screen and toolbar buttons in another PR.

File: game.py (hunk 2 above)
Comment (quozl) [reply game.py:107]: I agree, do not call pygame in `__init__`, call in `run`.

File: game.py (hunk 1 above)
Comment (quozl) [reply game.py:516]: Understandability trumps modularity.  This isn't how we do it everywhere else.  Please change the code so that the events are not _stored_ but are instead _processed_ within the loop that iterates over them, as before.

Comment (vaibhav-sangwan) [reply game.py]: [cfac48b](https://github.com/sugarlabs/ball-and-brick-activity/pull/24/commits/cfac48b4faf83e60c596f7fa06eab93d1d9399d4) resolves this.

**Comment by chimosky** [comment]
Reviewed and tested cfac48b, works as expected. Thank you!
